### DISCUSSION TITLE
SoftwareMock of ArduinoServer (emulate the telemetrix server part (which runs on an Arduino)?)

### DISCUSSION THREAD

**Discussion by esanya** [MrYsLab/telemetrix#31] [Ideas]
Hi,

I have started to implement a python program which will use telemetrix to control Arduino.

As I'm now in an early phase, I would implement / test the logic without an Arduino

My question / problem is:
-would it be possible somehow emulate the telemetrix server part (which runs on an Arduino)?
-I have tried to open a virtual serial port like: socat -d -d pty,raw,echo=0 tcp-listen:8082
-and check what is arriving there by: nc localhost 8082

Thanks,
Regards,
Sandor

**Comment by MrYsLab** [comment]
@esanya, I am not sure why you wish to see the data going across the serial link from the client to the server. That is only half of what the server does. Writing a full emulator would also need to generate reports from the server to the client.

However, there are a couple of possibly more straightforward solutions.

The first suggestion does not require additional hardware but just modifying the telemetrix file to bypass some hardware checks. Here is a [link to a file](https://gist.github.com/MrYsLab/700c3ee44d2e02b88560df482ee77414) that does just that.

When I run this file against the blink.py example, here is the output I get:
```
Telemetrix:  Version 1.10

Copyright (c) 2021 Alan Yorinks. All Rights Reserved.

Opening all potential serial ports...

Waiting 4 seconds(arduino_wait) for Arduino devices to reset...
b'\x01\x0f'
b'\x03\x01\x06\x01'
1
b'\x03\x02\x06\x01'
0
b'\x03\x02\x06\x00'
1
b'\x03\x02\x06\x01'
0
b'\x03\x02\x06\x00'
1
b'\x03\x02\x06\x01'
0
b'\x03\x02\x06\x00'
1
b'\x03\x02\x06\x01'
0
b'\x03\x02\x06\x00'
1
b'\x03\x02\x06\x01'
0
b'\x03\x02\x06\x00'
```
The 1's and zeroes being printed out from the blink example indicate the state of the output port.

You can also use an Arduino to loop back all commands sent to it.

You would run your code with an unmodified telemetrix.py file to do this. In the Telemetrix4Arduino.ino file, uncomment out [this line](https://github.com/MrYsLab/Telemetrix4Arduino/blob/fa7d30c9f42ccc3953a3e705df5d84ee29b04004/examples/Telemetrix4Arduino/Telemetrix4Arduino.ino#L1632) and [this one](https://github.com/MrYsLab/Telemetrix4Arduino/blob/fa7d30c9f42ccc3953a3e705df5d84ee29b04004/examples/Telemetrix4Arduino/Telemetrix4Arduino.ino#L1647)

I hope this was helpful.
